Support for domain specific metadata

ochiengolanga opened the pull request:
## What does this change?

Introduces the ability to define additional metadata referred to as domain specific metadata via configuration. The configuration will render additional domain specific metadata sections in the image metadata panel on the user interface and will be used by `metadata-editor` service to validate domain metadata updates via the metadata edit API.

**Note** : This PR introduces a new field [domainMetadata](https://github.com/bbc/grid/blob/684fcd301c22e89fe09c92d514225b532b075e4d/common-lib/src/main/scala/com/gu/mediaservice/model/ImageMetadata.scala#L29) in the ImageMetadata.scala case class and requires an update on the Elasticsearch mapping. This can be done using the `UpdateMapping` script. Refer to [UpdateMapping](https://github.com/guardian/grid/tree/main/scripts#updatemapping) documentation.

Link to proposal - https://github.com/guardian/grid/discussions/3441

## How can success be measured?

Adding the following configuration to common configuration should 

* Render "BBC Programmes" and "BBC Things" metadata collapsible sections and metadata  fields in the image metadata panel as shown in the screenshots below.

* Validate domain metadata update payload on `metadata-editor` service during PUT API request to update image metadata.

The rendered fields in each domain section should allow one to modify and save field values.

```hocon
domainMetadata.specifications = [
  {
    # Uniquely identifies a specification
    name = "bbcprogrammes"

    # Represents a human readable name for the specification. Used to render domain metadata collapsible header on the
    # user interface.
    label = "BBC Programmes"

    # Optional string that represents a human readable description of the specification. Rendered in the domain metadata
    # collapsible.
    description = "Model used for BBC programmes"

    # List of fields belonging to this particular specification.
    fields = [
      {
        # Uniquely identifies a field in a specification
        name = "programmeTitle"

        # Used to render field name and input field label on the user interface.
        label = "Title"

        # Represents the type of the field and it's value.
        # Valid / supported field type values are:
        #  - string - Represents text based values and renders a single-line text input field
        #  - integer - Represents numeric based values and renders numeric input field
        #  - datetime - Represents date and time values and renders a date and time input field
        #  - select - Represents text based values from controlled vocabularies and renders options as a drop-down list / select field.
        type = "string"

        # List of options rendered in the drop-down list / select field.
        # Mandatory for 'select' type.
        # options = ["Option 1", "Option 2"]
      }
      {
        name = "originalTxDate"
        label = "Original TX Date"
        type = "datetime"
      }
      {
        name = "episodeTxWeek"
        label = "Episode TX Week"
        type = "integer"
      }
      {
        name = "programmeGenre"
        label = "Genre"
        type = "select"
        options = ["Option 1", "Option 2"]
      }
    ]
  }
  {
    name = "bbcthings"
    label = "BBC Things"
    description = "Metadata for all things BBC Things"
    fields = [
      {
        name = "things"
        label = "Things"
        type = "string"
      }
    ]
  }
]
```

## Screenshots

### Domain metadata collapsibles on the image metadata panel

![Image metadata expanded - edited](https://user-images.githubusercontent.com/12212239/136617269-2b50b27c-f06f-4421-bfe5-83a9f238a0f3.png)

### Usage

![Image metadata - model demo](https://user-images.githubusercontent.com/12212239/136617289-2c6e2850-7229-4d0b-9c6f-485a92a53d21.gif)

## Who should look at this?

@guardian/digital-cms


## Tested? Documented?
- [X] locally by committer
- [ ] locally by Guardian reviewer
- [ ] on the Guardian's TEST environment
- [X] relevant documentation added or amended (if needed)

andrew-nowak reviewed (commented):
This looks really good! I'll defer to @paperboyo on UX considerations, and I have a small nitpick about namings in the config. 
Have you thought about what happens if the metadata specs need to change at all in the future? eg. what would happen if a field was added, deleted, or the type was changed? If there's any gotchas there it'd be great to document them.

ochiengolanga commented on docs/07-extending/05-domain-metadata.md:
```diff
@@ -0,0 +1,75 @@
+    # List of fields belonging to this particular specification.
+    fields = [
+      {
+        # Uniquely identifies a field in a specification
+        name = "field-a"
+
+        # Used to render field name and input field label on the user interface.
+        label = "Field A"
```
This is now resolved. The use of `name` and `label` in the configuration is now consistent. Thanks for spotting the inconsistency.

ochiengolanga commented:
> Hi. Thanks to Andrew I had an opportunity to have a play on TEST. I’m very, very excited about all the input options, they will come in handy later, too!
> 
> Just some mainly minor UI issues:
> 
> 1. I wonder: should the Show/Hide states of these (and Show additional metadata too) be kept in local storage similar to states of [locking the panels](https://github.com/guardian/grid/pull/2890)? I just imagine a person whose job is to read and fill these having to expand them for **every** image…
> 2. [minor] Is it possible to use Agate font in the dropdown menu (currently, it’s some system font)?
>    ![image](https://user-images.githubusercontent.com/6032869/139417370-d47e38e1-6ff6-4ad9-a120-8db5976514d4.png)
> 3. Unknown (empty) values should be rendered darker (like People below):
>    ![image](https://user-images.githubusercontent.com/6032869/139417622-e426d6e4-8de9-4652-9d96-62a787cece3e.png)

@paperboyo The suggestion to remember Show/Hide states for both the additional and domain metadata sections improves the UX. Thanks for the suggestion. 1,2 and 3 have been addressed in this [commit](https://github.com/guardian/grid/pull/3496/commits/1c02df065789aa4012d3a2b4e158aa258e7f8f82). Mind giving it a try?

ochiengolanga commented:
> This looks really good! I'll defer to @paperboyo on UX considerations, and I have a small nitpick about namings in the config. Have you thought about what happens if the metadata specs need to change at all in the future? eg. what would happen if a field was added, deleted, or the type was changed? If there's any gotchas there it'd be great to document them.

@andrew-nowak Yes, modification of the spec and the effects of the modification were taken into consideration. Despite the spec being more of a static config, modification is still allowed. From tests, here's what happens in each spec change scenario:

- Added field(s) to the spec will be rendered on the UI, used to validate the metadata update payload by the API and will be stored in the Edits DynamoDB table and Elasticsearch.
- Removed field(s) from the metadata spec will not be rendered on the UI even if it has a value, the field will be removed from the metadata update payload during validation and will not be stored in the Edits DynamoDB table and Elasticsearch.
- Modifying the field type will throws a format error on the UI (the gotcha) if the image metadata has a value (except `integer` to `string`, `string` to `select` and `select` to `string`). For example:	

> - Changing the type from `string` to `integer` the UI will throw "Error: [ngModel:numfmt] Expected `String` to be a number"

> - Changing the type from `string` to `datetime` the UI will throw "Error: [ngModel:datefmt] Expected `11` to be a date" if the string is not date format

> - Changing the type from `integer` to `datetime` the UI will throw "Error: [ngModel:datefmt] Expected `1` to be a date"

Despite these format errors, a user can still modify and save the value of the field using the new type.

Happy to receive any suggestions or thoughts on this.
